File: app/src/main/java/org/wikipedia/views/ObservableWebView.java
package org.wikipedia.views;

import android.content.Context;
import android.graphics.Canvas;
import android.os.SystemClock;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.ViewConfiguration;
import android.webkit.WebView;

import org.wikipedia.WikipediaApp;
import org.wikipedia.events.WebViewInvalidateEvent;
import org.wikipedia.util.DimenUtil;

import java.util.ArrayList;
import java.util.List;

public class ObservableWebView extends WebView {
    private static final WebViewInvalidateEvent INVALIDATE_EVENT = new WebViewInvalidateEvent();

    private List<OnClickListener> onClickListeners;
    private List<OnScrollChangeListener> onScrollChangeListeners;
    private List<OnDownMotionEventListener> onDownMotionEventListeners;
    private List<OnUpOrCancelMotionEventListener> onUpOrCancelMotionEventListeners;
    private List<OnContentHeightChangedListener> onContentHeightChangedListeners;
    private OnFastScrollListener onFastScrollListener;

    private int contentHeight = 0;
    private float touchStartX;
    private float touchStartY;
    private int touchSlop;

    private long lastScrollTime;
    private int totalAmountScrolled;

    /**
    * Threshold (in pixels) of continuous scrolling, to be considered "fast" scrolling.
    */
    private static final int FAST_SCROLL_THRESHOLD = (int) (1000 * DimenUtil.getDensityScalar());

    /**
    * Maximum single scroll amount (in pixels) to be considered a "human" scroll.
    * Otherwise it's probably a programmatic scroll, which we won't count.
    */
    private static final int MAX_HUMAN_SCROLL = (int) (500 * DimenUtil.getDensityScalar());

    /**
     * Maximum amount of time that needs to elapse before the previous scroll amount
     * is "forgotten." That is, if the user scrolls once, then scrolls again within this
     * time, then the two scroll actions will be added together as one, and counted towards
     * a possible "fast" scroll.
     */
    private static final int MAX_MILLIS_BETWEEN_SCROLLS = 500;

    public void addOnClickListener(OnClickListener onClickListener) {
        onClickListeners.add(onClickListener);
    }

    public void addOnScrollChangeListener(OnScrollChangeListener onScrollChangeListener) {
        onScrollChangeListeners.add(onScrollChangeListener);
    }

    public void addOnDownMotionEventListener(OnDownMotionEventListener onDownMotionEventListener) {
        onDownMotionEventListeners.add(onDownMotionEventListener);
    }

    public void addOnUpOrCancelMotionEventListener(OnUpOrCancelMotionEventListener onUpOrCancelMotionEventListener) {
        onUpOrCancelMotionEventListeners.add(onUpOrCancelMotionEventListener);
    }

    public void addOnContentHeightChangedListener(OnContentHeightChangedListener onContentHeightChangedListener) {
        onContentHeightChangedListeners.add(onContentHeightChangedListener);
    }

    public void setOnFastScrollListener(OnFastScrollListener onFastScrollListener) {
        this.onFastScrollListener = onFastScrollListener;
    }

    public void clearAllListeners() {
        onClickListeners.clear();
        onScrollChangeListeners.clear();
        onDownMotionEventListeners.clear();
        onUpOrCancelMotionEventListeners.clear();
        onContentHeightChangedListeners.clear();
        onFastScrollListener = null;
    }

    public interface OnClickListener {
        boolean onClick(float x, float y);
    }

    public interface OnScrollChangeListener {
        void onScrollChanged(int oldScrollY, int scrollY, boolean isHumanScroll);
    }

    public interface OnDownMotionEventListener {
        void onDownMotionEvent();
    }

    public interface OnUpOrCancelMotionEventListener {
        void onUpOrCancelMotionEvent();
    }

    public interface OnContentHeightChangedListener {
        void onContentHeightChanged(int contentHeight);
    }

    public interface OnFastScrollListener {
        void onFastScroll();
    }

    public void copyToClipboard() {
        // Simulate a Ctrl-C key press, which copies the current selection to the clipboard.
        // Seems to work across all APIs.
        dispatchKeyEvent(new KeyEvent(SystemClock.uptimeMillis(), SystemClock.uptimeMillis(),
                KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_C, 0, KeyEvent.META_CTRL_ON));
        dispatchKeyEvent(new KeyEvent(SystemClock.uptimeMillis(), SystemClock.uptimeMillis(),
                KeyEvent.ACTION_UP, KeyEvent.KEYCODE_C, 0, KeyEvent.META_CTRL_ON));
    }

    public ObservableWebView(Context context) {
        super(context);
        init();
    }

    public ObservableWebView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public ObservableWebView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
        onClickListeners = new ArrayList<>();
        onScrollChangeListeners = new ArrayList<>();
        onDownMotionEventListeners = new ArrayList<>();
        onUpOrCancelMotionEventListeners = new ArrayList<>();
        onContentHeightChangedListeners = new ArrayList<>();
        touchSlop = ViewConfiguration.get(getContext()).getScaledTouchSlop();
    }

    @Override
    protected void onScrollChanged(int left, int top, int oldLeft, int oldTop) {
        super.onScrollChanged(left, top, oldLeft, oldTop);
        boolean isHumanScroll = Math.abs(top - oldTop) < MAX_HUMAN_SCROLL;
        for (OnScrollChangeListener listener : onScrollChangeListeners) {
            listener.onScrollChanged(oldTop, top, isHumanScroll);
        }
        if (!isHumanScroll) {
            return;
        }
        totalAmountScrolled += (top - oldTop);
        if (Math.abs(totalAmountScrolled) > FAST_SCROLL_THRESHOLD
                && onFastScrollListener != null) {
            onFastScrollListener.onFastScroll();
            totalAmountScrolled = 0;
        }
        lastScrollTime = System.currentTimeMillis();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
                for (OnDownMotionEventListener listener : onDownMotionEventListeners) {
                    listener.onDownMotionEvent();
                }
                if (System.currentTimeMillis() - lastScrollTime > MAX_MILLIS_BETWEEN_SCROLLS) {
                    totalAmountScrolled = 0;
                }
                touchStartX = event.getX();
                touchStartY = event.getY();
                break;
            case MotionEvent.ACTION_UP:
                if (Math.abs(event.getX() - touchStartX) <= touchSlop
                        && Math.abs(event.getY() - touchStartY) <= touchSlop) {
                    for (OnClickListener listener : onClickListeners) {
                        if (listener.onClick(event.getX(), event.getY())) {
                            return true;
                        }
                    }
                }
            case MotionEvent.ACTION_CANCEL:
                for (OnUpOrCancelMotionEventListener listener : onUpOrCancelMotionEventListeners) {
                    listener.onUpOrCancelMotionEvent();
                }
                break;
            default:
                // Do nothing for all the other things
                break;
        }
        return super.onTouchEvent(event);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (isInEditMode()) {
            return;
        }
        if (contentHeight != getContentHeight()) {
            contentHeight = getContentHeight();
            for (OnContentHeightChangedListener listener : onContentHeightChangedListeners) {
                listener.onContentHeightChanged(contentHeight);
            }
        }
        WikipediaApp.getInstance().getBus().post(INVALIDATE_EVENT);
    }
}

File: app/src/main/java/org/wikipedia/views/ObservableWebView.kt
package org.wikipedia.views

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.os.SystemClock
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.ViewConfiguration
import android.webkit.WebView
import org.wikipedia.WikipediaApp
import org.wikipedia.events.WebViewInvalidateEvent
import org.wikipedia.util.DimenUtil.densityScalar
import org.wikipedia.views.FrameLayoutNavMenuTriggerer.Companion.setChildViewScrolled
import java.util.*
import kotlin.math.abs

class ObservableWebView : WebView {

    fun interface OnClickListener {
        fun onClick(x: Float, y: Float): Boolean
    }

    fun interface OnScrollChangeListener {
        fun onScrollChanged(oldScrollY: Int, scrollY: Int, isHumanScroll: Boolean)
    }

    fun interface OnDownMotionEventListener {
        fun onDownMotionEvent()
    }

    fun interface OnUpOrCancelMotionEventListener {
        fun onUpOrCancelMotionEvent()
    }

    fun interface OnContentHeightChangedListener {
        fun onContentHeightChanged(contentHeight: Int)
    }
    fun interface OnFastScrollListener {
        fun onFastScroll()
    }

    private var onClickListeners: MutableList<OnClickListener> = ArrayList()
    private var onScrollChangeListeners: MutableList<OnScrollChangeListener> = ArrayList()
    private var onDownMotionEventListeners: MutableList<OnDownMotionEventListener> = ArrayList()
    private var onUpOrCancelMotionEventListeners: MutableList<OnUpOrCancelMotionEventListener> = ArrayList()
    private var onContentHeightChangedListeners: MutableList<OnContentHeightChangedListener> = ArrayList()
    private var onFastScrollListener: OnFastScrollListener? = null
    private var currentContentHeight = 0
    private var lastScrollTime: Long = 0
    private var totalAmountScrolled = 0
    private var drawEventsWhileSwiping = 0
    private var touchSlop = ViewConfiguration.get(context).scaledTouchSlop
    var touchStartX = 0f
        private set
    var touchStartY = 0f
        private set
    var lastTouchX = 0f
        private set
    var lastTouchY = 0f
        private set

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet? = null) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int) : super(context, attrs, defStyle)

    fun addOnClickListener(onClickListener: OnClickListener) {
        onClickListeners.add(onClickListener)
    }

    fun addOnScrollChangeListener(onScrollChangeListener: OnScrollChangeListener) {
        onScrollChangeListeners.add(onScrollChangeListener)
    }

    fun addOnDownMotionEventListener(onDownMotionEventListener: OnDownMotionEventListener) {
        onDownMotionEventListeners.add(onDownMotionEventListener)
    }

    fun addOnUpOrCancelMotionEventListener(onUpOrCancelMotionEventListener: OnUpOrCancelMotionEventListener) {
        onUpOrCancelMotionEventListeners.add(onUpOrCancelMotionEventListener)
    }

    fun addOnContentHeightChangedListener(onContentHeightChangedListener: OnContentHeightChangedListener) {
        onContentHeightChangedListeners.add(onContentHeightChangedListener)
    }

    fun clearAllListeners() {
        onClickListeners.clear()
        onScrollChangeListeners.clear()
        onDownMotionEventListeners.clear()
        onUpOrCancelMotionEventListeners.clear()
        onContentHeightChangedListeners.clear()
        onFastScrollListener = null
    }

    override fun onScrollChanged(left: Int, top: Int, oldLeft: Int, oldTop: Int) {
        super.onScrollChanged(left, top, oldLeft, oldTop)
        val isHumanScroll = abs(top - oldTop) < MAX_HUMAN_SCROLL
        onScrollChangeListeners.forEach {
            it.onScrollChanged(oldTop, top, isHumanScroll)
        }
        if (!isHumanScroll) {
            return
        }
        totalAmountScrolled += top - oldTop
        if (abs(totalAmountScrolled) > FAST_SCROLL_THRESHOLD &&
                onFastScrollListener != null) {
            onFastScrollListener!!.onFastScroll()
            totalAmountScrolled = 0
        }
        lastScrollTime = System.currentTimeMillis()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (event.buttonState == MotionEvent.BUTTON_SECONDARY &&
                event.actionMasked == MotionEvent.ACTION_DOWN) {
            handleMouseRightClick(event.x, event.y)
            return true
        }
        lastTouchX = event.x
        lastTouchY = event.y
        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                onDownMotionEventListeners.forEach {
                    it.onDownMotionEvent()
                }
                if (System.currentTimeMillis() - lastScrollTime > MAX_MILLIS_BETWEEN_SCROLLS) {
                    totalAmountScrolled = 0
                }
                touchStartX = event.x
                touchStartY = event.y
                drawEventsWhileSwiping = 0
            }
            MotionEvent.ACTION_UP -> {
                if (abs(event.x - touchStartX) <= touchSlop &&
                        abs(event.y - touchStartY) <= touchSlop) {
                    if (hitTestResult.type == HitTestResult.UNKNOWN_TYPE) {
                        if (onClickListeners.any { it.onClick(event.x, event.y) }) {
                            return true
                        }
                    }
                }
                drawEventsWhileSwiping = 0
                onUpOrCancelMotionEventListeners.forEach {
                    it.onUpOrCancelMotionEvent()
                }
                drawEventsWhileSwiping = 0
            }
            MotionEvent.ACTION_CANCEL -> {
                onUpOrCancelMotionEventListeners.forEach {
                    it.onUpOrCancelMotionEvent()
                }
                drawEventsWhileSwiping = 0
            }
            else -> { }
        }
        return super.onTouchEvent(event)
    }

    private fun handleMouseRightClick(x: Float, y: Float) {
        val eventTimeTravelMillis = 1000
        post {
            dispatchTouchEvent(MotionEvent.obtain(SystemClock.uptimeMillis(),
                    SystemClock.uptimeMillis() + eventTimeTravelMillis,
                    MotionEvent.ACTION_DOWN, x, y, 0))
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (isInEditMode) {
            return
        }
        drawEventsWhileSwiping++
        if (drawEventsWhileSwiping > SWIPE_DRAW_TOLERANCE) {
            setChildViewScrolled()
        }
        if (currentContentHeight != contentHeight) {
            currentContentHeight = contentHeight
            onContentHeightChangedListeners.forEach {
                it.onContentHeightChanged(currentContentHeight)
            }
        }
        WikipediaApp.getInstance().bus.post(INVALIDATE_EVENT)
    }

    companion object {
        private val INVALIDATE_EVENT = WebViewInvalidateEvent()
        private val FAST_SCROLL_THRESHOLD = (1000 * densityScalar).toInt()
        private val MAX_HUMAN_SCROLL = (500 * densityScalar).toInt()
        private const val MAX_MILLIS_BETWEEN_SCROLLS = 500
        private const val SWIPE_DRAW_TOLERANCE = 4
    }
}
